File: src/nfan/fan/sys/Month.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   23 Feb 07  Andy Frank  Creation
//

namespace Fan.Sys
{
  /// <summary>
  /// Month.
  /// </summary>
  public sealed class Month : Enum
  {

    public static readonly Month m_jan = new Month(0,  "jan");
    public static readonly Month m_feb = new Month(1,  "feb");
    public static readonly Month m_mar = new Month(2,  "mar");
    public static readonly Month m_apr = new Month(3,  "apr");
    public static readonly Month m_may = new Month(4,  "may");
    public static readonly Month m_jun = new Month(5,  "jun");
    public static readonly Month m_jul = new Month(6,  "jul");
    public static readonly Month m_aug = new Month(7,  "aug");
    public static readonly Month m_sep = new Month(8,  "sep");
    public static readonly Month m_oct = new Month(9,  "oct");
    public static readonly Month m_nov = new Month(10, "nov");
    public static readonly Month m_dec = new Month(11, "dec");

    internal static readonly Month[] array =
    {
      m_jan, m_feb, m_mar, m_apr, m_may, m_jun,
      m_jul, m_aug, m_sep, m_oct, m_nov, m_dec
    };

    public static readonly List m_values = new List(Sys.MonthType, array).ro();

    private Month(int ordinal, string name)
    {
      Enum.make_(this, Int.m_pos[ordinal], Str.make(name).intern());
      this.ord = ordinal;
      this.localeAbbrKey = Str.make(name + "Abbr");
      this.localeFullKey = Str.make(name + "Full");
    }

    public static Month fromStr(Str name) { return fromStr(name, Bool.True); }
    public static Month fromStr(Str name, Bool check)
    {
      return (Month)doFromStr(Sys.MonthType, name, check);
    }

    public override Type type() { return Sys.MonthType; }

    public Month increment() { return array[(ord+1)%array.Length]; }

    public Month decrement() { return ord == 0 ? array[array.Length-1] : array[ord-1]; }

    public Int numDays(Int year)
    {
      if (DateTime.isLeapYear((int)year.val))
        return Int.m_pos[DateTime.daysInMonLeap[ord]];
      else
        return Int.m_pos[DateTime.daysInMon[ord]];
    }

    public Str toLocale() { return toLocale(null); }
    public Str toLocale(Str pattern)
    {
      if (pattern == null) return localeAbbr();
      if (pattern.isEveryChar('M'))
      {
        switch (pattern.val.Length)
        {
          case 1: return Str.make(""+(ord+1));
          case 2: return Str.make(ord < 9 ? "0" + (ord+1) : ""+(ord+1));
          case 3: return localeAbbr();
          case 4: return localeFull();
        }
      }
      throw ArgErr.make("Invalid pattern: " + pattern).val;
    }

    public Str localeAbbr() { return abbr(Locale.current()); }
    public Str abbr(Locale locale)
    {
      return locale.get(Str.sysStr, localeAbbrKey);
    }

    public Str localeFull() { return full(Locale.current()); }
    public Str full(Locale locale)
    {
      return locale.get(Str.sysStr, localeFullKey);
    }

    internal readonly int ord;
    readonly Str localeAbbrKey;
    readonly Str localeFullKey;
  }
}

File: src/webappClient/javascript/sys/Month.js
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Feb 09  Andy Frank  Creation
//

/**
 * Month
 */
var sys_Month = sys_Enum.extend(
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  $ctor: function(ordinal, name)
  {
    this.$make(ordinal, name);
  },

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  increment: function()
  {
    var arr = sys_Month.values;
    return arr[(this.m_ordinal+1) % arr.length];
  },

  decrement: function()
  {
    var arr = sys_Month.values;
    return this.m_ordinal == 0 ? arr[arr.length-1] : arr[this.m_ordinal-1];
  },

  /*
  public long numDays(long year)
  {
    if (DateTime.isLeapYear((int)year))
      return DateTime.daysInMonLeap[ord];
    else
      return DateTime.daysInMon[ord];
  }
  */

  type: function()
  {
    return sys_Type.find("sys::Month");
  },

});

//////////////////////////////////////////////////////////////////////////
// Static Methods
//////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////
// Static Fields
//////////////////////////////////////////////////////////////////////////

sys_Month.jan = new sys_Month(0,  "jan");
sys_Month.feb = new sys_Month(1,  "feb");
sys_Month.mar = new sys_Month(2,  "mar");
sys_Month.apr = new sys_Month(3,  "apr");
sys_Month.may = new sys_Month(4,  "may");
sys_Month.jun = new sys_Month(5,  "jun");
sys_Month.jul = new sys_Month(6,  "jul");
sys_Month.aug = new sys_Month(7,  "aug");
sys_Month.sep = new sys_Month(8,  "sep");
sys_Month.oct = new sys_Month(9,  "oct");
sys_Month.nov = new sys_Month(10, "nov");
sys_Month.dec = new sys_Month(11, "dec");

sys_Month.values =
[
  sys_Month.jan, sys_Month.feb, sys_Month.mar, sys_Month.apr, sys_Month.may, sys_Month.jun,
  sys_Month.jul, sys_Month.aug, sys_Month.sep, sys_Month.oct, sys_Month.nov, sys_Month.dec
];